src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImpl.scala

```diff
@@ -1,133 +1,85 @@
 
-import javax.inject.Named
+import javax.inject.Inject
 
-import akka.actor.ActorRef
-import akka.event.EventStream
-import com.google.inject.Inject
-import mesosphere.marathon.MarathonSchedulerActor.ScaleApp
-import mesosphere.marathon.MarathonSchedulerDriverHolder
 import mesosphere.marathon.Protos.MarathonTask
-import mesosphere.marathon.core.launchqueue.LaunchQueue
-import mesosphere.marathon.core.task.bus.MarathonTaskStatus.WithMesosStatus
-import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskStatusUpdate
-import mesosphere.marathon.core.task.bus.{ TaskStatusEmitter, TaskStatusObservables }
-import mesosphere.marathon.core.task.tracker.TaskStatusUpdateProcessor
-import mesosphere.marathon.event.{ EventModule, MesosStatusUpdateEvent }
-import mesosphere.marathon.health.HealthCheckManager
-import mesosphere.marathon.state.{ PathId, AppRepository, Timestamp }
+import mesosphere.marathon.core.base.Clock
+import mesosphere.marathon.core.task.tracker.{ TaskStatusUpdateStep, TaskStatusUpdateProcessor }
+import mesosphere.marathon.metrics.Metrics.Timer
+import mesosphere.marathon.metrics.{ MetricPrefixes, Metrics }
+import mesosphere.marathon.state.{ PathId, AppDefinition, AppRepository, Timestamp }
 import mesosphere.marathon.tasks.{ TaskIdUtil, TaskTracker }
-import org.apache.mesos.Protos.{ TaskState, TaskStatus }
+import org.apache.mesos.Protos.TaskStatus
 import org.slf4j.LoggerFactory
 
 import scala.concurrent.Future
-import scala.util.control.NonFatal
 
 /**
-  * Processes task status update events, mostly to update the task tracker.
+  * Executes the given TaskStatusUpdateSteps for every update.
   */
 class TaskStatusUpdateProcessorImpl @Inject() (
-    taskStatusEmitter: TaskStatusEmitter,
-    appRepository: AppRepository,
-    launchQueue: LaunchQueue,
-    @Named(EventModule.busName) eventBus: EventStream,
-    @Named("schedulerActor") schedulerActor: ActorRef,
+    metrics: Metrics,
+    clock: Clock,
     taskIdUtil: TaskIdUtil,
-    healthCheckManager: HealthCheckManager,
+    appRepository: AppRepository,
     taskTracker: TaskTracker,
-    marathonSchedulerDriverHolder: MarathonSchedulerDriverHolder) extends TaskStatusUpdateProcessor {
+    steps: Seq[TaskStatusUpdateStep]) extends TaskStatusUpdateProcessor {
+  import scala.concurrent.ExecutionContext.Implicits.global
 
   private[this] val log = LoggerFactory.getLogger(getClass)
 
-  //TODO: fix style issue and enable this scalastyle check
-  //scalastyle:off cyclomatic.complexity method.length
-  def publish(incoming: TaskStatusObservables.TaskStatusUpdate): Future[Unit] = incoming match {
-    case TaskStatusUpdate(timestamp, taskId, WithMesosStatus(status)) =>
-      val appId = taskIdUtil.appId(taskId)
+  private[this] val publishFutureTimer: Timer =
+    metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "publishFuture"))
+  private[this] val fetchTaskTimer: Timer = metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "fetchTask"))
```

Comment:
I wanted to provide a metric from the caller perspective. How much time does it need in the task status update? It might be different from other call sites because of caching, flow, etc.

I can remove it, if you want.